pkg/controller/history/controller_history.go

```diff
@@ -0,0 +1,487 @@
+		if errors.IsAlreadyExists(err) {
+			probe++
+			continue
+		}
+		return created, err
+	}
+}
+
+func (rh *realHistory) UpdateControllerRevision(revision *apps.ControllerRevision, newRevision int64) (*apps.ControllerRevision, error) {
+	obj, err := scheme.Scheme.DeepCopy(revision)
+	if err != nil {
+		return nil, err
+	}
+	clone := obj.(*apps.ControllerRevision)
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		if clone.Revision == newRevision {
+			return nil
+		}
+		clone.Revision = newRevision
+		_, updateErr := rh.client.Apps().ControllerRevisions(clone.Namespace).Update(clone)
+		if updateErr == nil {
+			return nil
+		}
+		if updated, err := rh.lister.ControllerRevisions(clone.Namespace).Get(clone.Name); err == nil {
+			// make a copy so we don't mutate the shared cache
+			obj, err := scheme.Scheme.DeepCopy(updated)
+			if err != nil {
+				return err
+			}
+			clone = obj.(*apps.ControllerRevision)
+		}
+		return updateErr
+	})
+	return clone, err
+}
+
+func (rh *realHistory) DeleteControllerRevision(revision *apps.ControllerRevision) error {
+	return rh.client.Apps().ControllerRevisions(revision.Namespace).Delete(revision.Name, nil)
+}
+
+func (rh *realHistory) AdoptControllerRevision(parent metav1.Object, parentKind schema.GroupVersionKind, revision *apps.ControllerRevision) (*apps.ControllerRevision, error) {
+	// Return an error if the parent does not own the revision
+	if owner := controller.GetControllerOf(revision); owner != nil {
+		return nil, fmt.Errorf("attempt to adopt revision owned by %v", owner)
+	}
+	// Use strategic merge patch to add an owner reference indicating a controller ref
+	return rh.client.Apps().ControllerRevisions(parent.GetNamespace()).Patch(revision.GetName(),
+		types.StrategicMergePatchType, []byte(fmt.Sprintf(
+			`{"metadata":{"ownerReferences":[{"apiVersion":"%s","kind":"%s","name":"%s","uid":"%s","controller":true,"blockOwnerDeletion":true}],"uid":"%s"}}`,
+			parentKind.GroupVersion().String(), parentKind.Kind,
+			parent.GetName(), parent.GetUID(), revision.UID)))
+}
+
+func (rh *realHistory) ReleaseControllerRevision(parent metav1.Object, revision *apps.ControllerRevision) (*apps.ControllerRevision, error) {
+	// Use strategic merge patch to add an owner reference indicating a controller ref
+	released, err := rh.client.Apps().ControllerRevisions(revision.GetNamespace()).Patch(revision.GetName(),
+		types.StrategicMergePatchType,
+		[]byte(fmt.Sprintf(`{"metadata":{"ownerReferences":[{"$patch":"delete","uid":"%s"}],"uid":"%s"}}`, parent.GetUID(), revision.UID)))
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// We ignore deleted revisions
+			return nil, nil
+		}
+		if errors.IsInvalid(err) {
+			// We ignore cases where the parent no longer owns the revision or where the revision has no
+			// owner.
+			return nil, nil
+		}
+	}
+	return released, err
+}
+
+type fakeHistory struct {
```

Comment:
Since this is implementing an interface, can't it live side-by-side in testing code?